Add profiling information for knn vector queries

jdconrad opened the pull request:
This adds timers to the dfs phase to profile a knn vector query and provide a breakdown of several parts of the query. The following is an example of the current output:

```
{
...
        "dfs" : {
          "query": {
            "type" : "KnnVectorQuery",
            "description" : "KnnVectorQuery:image-vector[-5.0,...][100]",
            "time_in_nanos" : 1147131,
            "breakdown" : {
              "create_weight_count" : 1,
              "create_weight" : 18195,
              "rewrite_count" : 1,
              "rewrite" : 185796,
              "collector" : 4801
            }
          }
        },
...
```

Design decisions:
1. This change attempts to be minimally invasive, so no new classes, and we use an existing `ProfileResult` to capture the profiling results.
2. We add the dfs `ProfileResult` to the `SearchProfileQueryPhaseResult` as part of the response during the dfs query phase. This minimizes the change overall as the dfs ProfileResult piggybacks along with the regular query results. This is an alternative to either passing an extra set of profile results along through the next-phase chain or through the `QueryPhaseResultConsumer`. Both of these options would add significantly more code.
3. We don't need to make changes to the `SearchPhaseController` since the results are naturally pulled along with the regular query results.
4. We add two additional overrides to the existing `IndexSearcher` used as part of dfs so we can add timers to both `rewrite` and `createWeight`.

jdconrad committed:
add profiling for knn search in dfs phase

jdconrad committed:
dfs results for non-scroll queries

jdconrad committed:
move dfs profile results to query results

jdconrad committed:
checkpoint

jdconrad committed:
add detail for the knn breakdown

jdconrad committed:
remove dead files

jdconrad committed:
remove comment

jdconrad committed:
revert spc

jdconrad committed:
minimize code change

jdconrad committed:
Merge branch 'main' into profile

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @jdconrad, I've created a changelog YAML for you.

jdconrad committed:
Update docs/changelog/90200.yaml

jdconrad commented:
This relates to (https://github.com/elastic/elasticsearch/issues/89713).

jdconrad committed:
Merge branch 'main' into profile

jdconrad commented:
@elasticsearchmachine run elasticsearch-ci/bwc

jdconrad commented:
@elasticsearchmachine elasticsearch-ci/part-1

jdconrad commented:
@elasticsearchmachine run elasticsearch-ci/part-1

jdconrad committed:
Merge branch 'main' into profile

jdconrad committed:
update stream versions to 8_6_0

nik9000 commented on server/src/main/java/org/elasticsearch/search/dfs/DfsSearchResult.java:
```diff
@@ -33,6 +34,7 @@ public class DfsSearchResult extends SearchPhaseResult {
     private Map<String, CollectionStatistics> fieldStatistics = new HashMap<>();
     private DfsKnnResults knnResults;
     private int maxDoc;
+    private ProfileResult profileResult;
```
:+1:

nik9000 commented on server/src/main/java/org/elasticsearch/search/profile/SearchProfileQueryPhaseResult.java:
```diff
@@ -25,6 +26,8 @@
  */
 public class SearchProfileQueryPhaseResult implements Writeable {
 
+    private ProfileResult dfsProfileResult;
```
If we wanted to measure the `termStatistics` calls, for example, we'd replace this with a `DfsProfileShardResult`, right?

nik9000 reviewed (commented):
That's much less painful than I thought it'd be!

jdconrad commented on server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java:
```diff
@@ -39,6 +45,11 @@ public class DfsPhase {
 
     public void execute(SearchContext context) {
         try {
+            final boolean[] profile = new boolean[2];
```
I agree that a comment is needed. I can probably rework this to just a single boolean after thinking about it. The idea is to avoid tracking time for createWeight used for collecting term stats, and only turning on profiling for createWeight when run the actual vector query.

jdconrad replied:
@nik9000 maybe it would make sense to create a new `IndexSearcher` that wraps the context reader specifically for knn search instead of re-using the one designed for term stats? I just wasn't sure how heavy-weight it is to make a second one? I think having a separate one for term stats and a separate one for knn may make sense though? Probably also separate out the terms stats collection and knn search into separate private methods?

nik9000 replied:
Yeah. I'm not really sure the "right" way to fix it, but I'll bet there is some nicer way.....

jdconrad replied:
@nik9000 I updated this section to split out the dfs phase into two private methods where one does stats collection and the other does the knn search. Each method creates its own `IndexSearcher` to wrap the existing `IndexReader`. This allows us to override methods only relevant to profiling the knn search in what I consider more obvious way and eliminates the `boolean[]` hack. What do you think?

jdconrad committed:
add tests and split out dfs phase

jdconrad commented:
I added a few tests cases to 370_profile.yml as requested.

nik9000 commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/370_profile.yml:
```diff
@@ -161,3 +161,80 @@ disabling stored fields removes fetch sub phases:
   - is_false: profile.shards.0.fetch.children
+
+---
+dfs knn vector profiling:
```
Can you profile with the dfs phase enabled without knn? We shouldn't see anything, but I want to be extra paranoia that it runs and doesn't crash.

nik9000 replied:
I see you did it below but I didn't realize because of the name.

jdconrad replied:
I'll split these up.

nik9000 commented on server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java:
```diff
@@ -39,71 +45,150 @@ public class DfsPhase {
-                KnnVectorQueryBuilder knnVectorQueryBuilder = knnSearch.toQueryBuilder();
-                if (context.request().getAliasFilter().getQueryBuilder() != null) {
-                    knnVectorQueryBuilder.addFilterQuery(context.request().getAliasFilter().getQueryBuilder());
+        IndexSearcher searcher = new IndexSearcher(context.searcher().getIndexReader()) {
```
should this be `profile ? <your subclass> : context.searcher()`?

jtibshirani replied:
It'd be nice to add a comment here explaining why we can't just use the context searcher in both cases. I assume it's because we don't want to use the `QueryProfiler` that it uses?

jdconrad commented:
I chatted with @nik9000 and as a result of that conversation I'm going to do the following:
1. I'm going to try to break up the knn search branching a bit more so that profiling isn't so intertwined
2. I'm going to split up the yml profiling tests so it's clear what each particular search is testing

jdconrad committed:
response to pr comments

jdconrad committed:
Merge branch 'main' into profile

jdconrad commented:
@nik9000 I split the knn search into profiling and non-profiling methods. This seems cleaner. I also split up the 370_profile tests to make them easier to see what they're testing. I think this is ready for another pass.

jdconrad committed:
always set dfs profile result even when null

jdconrad committed:
add safety check

jdconrad commented:
@elasticsearchmachine run elasticsearch-ci/part-2

jtibshirani commented on server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java:
```diff
@@ -39,71 +44,162 @@ public class DfsPhase {
+            long knnTotalTime = System.nanoTime() - knnStartTime;
+            context.dfsResult()
+                .profileResult(
+                    new ProfileResult(
```
Small comment, maybe it'd be nice to have a static method on `ProfileResult` like `createKnnResult`? It could hide some of these very profiler-specific details like creating a map the right keys, etc.

jdconrad replied:
@jtibshirani I agree that it should be moved, but I don't know if ProfileResult is the correct place to put it. If we're going to follow up this up with profiling for stats, would it make more sense to add this to part of the follow up and have a full custom `DfsProfileResult`? Or I could add one for this PR. What do you think?

jtibshirani replied:
Sure, sounds good to do it in a follow-up.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java:
```diff
@@ -39,71 +44,162 @@ public class DfsPhase {
+        IndexSearcher searcher = new IndexSearcher(context.searcher().getIndexReader()) {
+            @Override
+            public Query rewrite(Query original) throws IOException {
+                rewriteTimer.start();
```
Maybe we can check for cancellation here like we do when gathering term statistics? It doesn't seem critical, but is nice for consistency.

jdconrad replied:
Good catch, I added these.

jtibshirani reviewed (commented):
Thanks @jdconrad! This looks good to me overall.

jdconrad committed:
Merge branch 'main' into profile

jdconrad committed:
response to pr feedback

jdconrad commented:
@jtibshirani Thank you for the review. I added an additional comment about the wrapper for the profiling path.

jdconrad committed:
update based on pr feedback

jdconrad committed:
Merge branch 'main' into profile

jdconrad commented:
@jtibshirani I updated the PR based on your feedback with the following:
1. Use the existing `QueryProfiler` to measure the vector query timings
2. Added an additional dfs profiling class to support serialization beyond what `ProfileResult` provides
3. Merged knn search back into a single path within `DfsPhase`

```
...
"dfs" : {
          "knn" : {
            "query" : [
              {
                "type" : "DocAndScoreQuery",
                "description" : "DocAndScore[100]",
                "time_in_nanos" : 445529,
                "breakdown" : {
                  "set_min_competitive_score_count" : 0,
                  "match_count" : 0,
                  "shallow_advance_count" : 0,
                  "set_min_competitive_score" : 0,
                  "next_doc" : 1786,
                  "match" : 0,
                  "next_doc_count" : 3,
                  "score_count" : 3,
                  "compute_max_score_count" : 0,
                  "compute_max_score" : 0,
                  "advance" : 4626,
                  "advance_count" : 1,
                  "score" : 2278,
                  "build_scorer_count" : 2,
                  "create_weight" : 128344,
                  "shallow_advance" : 0,
                  "create_weight_count" : 1,
                  "build_scorer" : 308495
                }
              }
            ],
            "rewrite_time" : 1287496,
            "collector" : [
              {
                "name" : "SimpleTopScoreDocCollector",
                "reason" : "search_top_hits",
                "time_in_nanos" : 19862
              }
            ]
          }
        },
...
```

jtibshirani reviewed (approved):
Thanks for the update, I like how we don't need any custom profiling logic now inside `DfsPhase`.

jdconrad commented:
@jtibshirani Thanks for the reviews! I agree this feels cleaner w/o the custom logic. I'll merge as soon as CI passes.
